packages/eslint-plugin/src/rules/no-unnecessary-condition.ts

```diff
@@ -219,12 +230,30 @@ export default createRule<Options, MessageId>({
     function checkIfBinaryExpressionIsNecessaryConditional(
       node: TSESTree.BinaryExpression,
     ): void {
-      if (
-        BOOL_OPERATORS.has(node.operator) &&
-        isLiteral(getNodeType(node.left)) &&
-        isLiteral(getNodeType(node.right))
-      ) {
+      if (!BOOL_OPERATORS.has(node.operator)) {
+        return;
+      }
+      const leftType = getNodeType(node.left);
+      const rightType = getNodeType(node.right);
+      if (isLiteral(leftType) && isLiteral(rightType)) {
         context.report({ node, messageId: 'literalBooleanExpression' });
+        return;
+      }
+      // Workaround for https://github.com/microsoft/TypeScript/issues/37160
+      if (isStrictCompilerOptionEnabled(compilerOptions, 'strictNullChecks')) {
+        const UNDEFINED = ts.TypeFlags.Undefined;
+        const NULL = ts.TypeFlags.Null;
+        if (
+          (leftType.flags === UNDEFINED &&
+            !typeContainsFlag(rightType, UNDEFINED)) ||
+          (rightType.flags === UNDEFINED &&
+            !typeContainsFlag(leftType, UNDEFINED)) ||
+          (leftType.flags === NULL && !typeContainsFlag(rightType, NULL)) ||
+          (rightType.flags === NULL && !typeContainsFlag(leftType, NULL))
+        ) {
+          context.report({ node, messageId: 'noOverlapBooleanExpression' });
+          return;
+        }
```

Comment:
I think this check is already in `no-unnecessary-type-assertion` (in the non-null assertion path).

https://github.com/typescript-eslint/typescript-eslint/blob/3c5659b1984c6eeb07e4e06d9a213d0daca96e48/packages/eslint-plugin/src/rules/no-unnecessary-type-assertion.ts#L203-L241

If I'm correct, could you share the code?